### ISSUE TITLE
Modell works until 20 mpi ranks, but Heap corruption above that

### ISSUE THREAD

**Issue by tandreasr** [MODFLOW-USGS/modflow6#1964]
Hi Martijn,

this is probably another issue for you :-)
And sorry, but I've found no way to add the parallel label from my Github account...

Please have a look at the attached model.
When running the latest development version of parallel modflow6 ([20240715 development build]) on Windows 11 like this:

   "E:\Modflow6\mf6.6.0.dev0_win64par\bin\mpiexec.exe" -np 24 "E:\Modflow6\mf6.6.0.dev0_win64par\bin\mf6.exe" -P

it produces a heap corruption error...
Below 22 mpi ranks all seems to work fine and up from -np 22 it always produces that error.
I'm running on a 24 core threadripper.
These are the tests I've done so far:

np    time   status
0	106s    ok   (serial version)
1	112s	ok
2	46s	ok
12	8s	ok
17	7s	ok
18	7s	ok
19	7s	ok
20	7s	ok
21	7s	ok
22		Heap corruption
23		Heap corruption
24		Heap corruption

I hope you're able to reproduce thar error.
In case of any additional questions just contact me :-)

Best regards
Andreas
[__usgs.zip](https://github.com/user-attachments/files/16391729/__usgs.zip)

**Comment by tandreasr** [comment]
Just an addition.
Using the latest nightly build from (20240730) the error still persists.

**Comment by mjr-deltares** [comment]
Thanks for reporting @tandreasr , we will have a look at it.

**Comment by jdhughes-usgs** [comment]
@mjr-deltares Cleaned up files that I ran on Denali.

[tandreasr.zip](https://github.com/user-attachments/files/16431204/tandreasr.zip)

The model runs in serial and with PETSC on a single processor. However, when running on 24 processors the following error occurs (last 2 rank errors):

```
[20]PETSC ERROR: #1 User provided function() line 0 in  unknown file
Rank 20 [Tue Jul 30 11:55:36 2024] [c0-0c0s6n3] application called MPI_Abort(MPI_COMM_WORLD, 50176059) - process 20
forrtl: error (76): Abort trap signal
Image              PC                Routine            Line        Source
mf6                0000000020482EBB  Unknown               Unknown  Unknown
libpthread-2.26.s  00007FFFF40FA2D0  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D65420  gsignal               Unknown  Unknown
libc-2.26.so       00007FFFF3D66A01  abort                 Unknown  Unknown
libmpich_intel.so  00007FFFF51D0868  Unknown               Unknown  Unknown
libmpich_intel.so  00007FFFF50DCBF6  MPI_Abort             Unknown  Unknown
libcraypetsc_inte  00007FFFF6034245  PetscSignalHandle     Unknown  Unknown
libcraypetsc_inte  00007FFFF6033A5D  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D654A0  Unknown               Unknown  Unknown
libmpich_intel.so  00007FFFF51E2FC0  Unknown               Unknown  Unknown
libmpich_intel.so  00007FFFF51E7808  MPIDI_CH3I_Progre     Unknown  Unknown
libmpich_intel.so  00007FFFF512D42D  MPIC_Wait             Unknown  Unknown
libmpich_intel.so  00007FFFF512DBAD  MPIC_Sendrecv         Unknown  Unknown
libmpich_intel.so  00007FFFF507AFCF  MPIR_Allgather_in     Unknown  Unknown
libmpich_intel.so  00007FFFF507B96D  MPIR_Allgather        Unknown  Unknown
libmpich_intel.so  00007FFFF514EF2A  MPIR_CRAY_Allgath     Unknown  Unknown
libmpich_intel.so  00007FFFF507BA68  MPIR_Allgather_im     Unknown  Unknown
libmpich_intel.so  00007FFFF507CA26  MPI_Allgather         Unknown  Unknown
libcraypetsc_inte  00007FFFF62AF450  PetscLayoutSetUp      Unknown  Unknown
libcraypetsc_inte  00007FFFF649DE3B  MatMPIAIJSetPreal     Unknown  Unknown
libcraypetsc_inte  00007FFFF64970D2  MatMPIAIJSetPreal     Unknown  Unknown
libcraypetsc_inte  00007FFFF649E539  MatCreateMPIAIJWi     Unknown  Unknown
libcraypetsc_inte  00007FFFF6510262  matcreatempiaijwi     Unknown  Unknown
mf6                00000000204454E0  petscmatrixmodule         108  PetscMatrix.F90
mf6                00000000202E76D5  numericalsolution        2364  NumericalSolution.f90
mf6                00000000202D858D  numericalsolution         504  NumericalSolution.f90
mf6                000000002001201B  mf6coremodule_mp_         361  mf6core.f90
mf6                0000000020012C8E  mf6coremodule_mp_          92  mf6core.f90
mf6                000000002000E832  mf6coremodule_mp_          43  mf6core.f90
mf6                000000002000E7FD  MAIN__                     11  mf6.f90
mf6                000000002000E792  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D5034A  __libc_start_main     Unknown  Unknown
mf6                000000002000E6AA  Unknown               Unknown  Unknown
cray/pe/tpsl/20.03.1/INTEL/19.0/x86_skylake/include --with-ptscotch-lib="-L/opt/cray/pe/tpsl/20.03.1/INTEL/19.0/x86_skylake/lib -lptscotch -lscotch -lptscotcherr -l
[22]PETSC ERROR: #1 User provided function() line 0 in  unknown file
Rank 22 [Tue Jul 30 11:55:36 2024] [c0-0c0s6n3] application called MPI_Abort(MPI_COMM_WORLD, 50176059) - process 22
forrtl: error (76): Abort trap signal
Image              PC                Routine            Line        Source
mf6                0000000020482EBB  Unknown               Unknown  Unknown
libpthread-2.26.s  00007FFFF40FA2D0  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D65420  gsignal               Unknown  Unknown
libc-2.26.so       00007FFFF3D66A01  abort                 Unknown  Unknown
libmpich_intel.so  00007FFFF51D0868  Unknown               Unknown  Unknown
libmpich_intel.so  00007FFFF50DCBF6  MPI_Abort             Unknown  Unknown
libcraypetsc_inte  00007FFFF6034245  PetscSignalHandle     Unknown  Unknown
libcraypetsc_inte  00007FFFF6033A5D  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D654A0  Unknown               Unknown  Unknown
libmpich_intel.so  00007FFFF51E2F30  Unknown               Unknown  Unknown
libmpich_intel.so  00007FFFF51E7808  MPIDI_CH3I_Progre     Unknown  Unknown
libmpich_intel.so  00007FFFF50F155D  MPIR_Waitall_impl     Unknown  Unknown
libmpich_intel.so  00007FFFF50F1D76  MPI_Waitall           Unknown  Unknown
libmpich_intel.so  00007FFFF5122196  pmpi_waitall          Unknown  Unknown
mf6                0000000020432A54  mpiroutermodule_m         325  MpiRouter.f90
mf6                0000000020431631  mpiroutermodule_m         217  MpiRouter.f90
mf6                000000002002D61B  virtualdatamanage         359  VirtualDataManager.f90
mf6                0000000020019384  runcontrolmodule_         160  RunControl.f90
mf6                00000000202E7477  numericalsolution        2353  NumericalSolution.f90
mf6                00000000202D858D  numericalsolution         504  NumericalSolution.f90
mf6                000000002001201B  mf6coremodule_mp_         361  mf6core.f90
mf6                0000000020012C8E  mf6coremodule_mp_          92  mf6core.f90
mf6                000000002000E832  mf6coremodule_mp_          43  mf6core.f90
mf6                000000002000E7FD  MAIN__                     11  mf6.f90
mf6                000000002000E792  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D5034A  __libc_start_main     Unknown  Unknown
mf6                000000002000E6AA  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D5034A  __libc_start_main     Unknown  Unknown
mf6                000000002000E6AA  Unknown               Unknown  Unknown
libc-2.26.so       00007FFFF3D5034A  __libc_start_main     Unknown  Unknown
mf6                000000002000E6AA  Unknown               Unknown  Unknown

```

**Comment by jdhughes-usgs** [comment]
@tandreasr I extracted the model extents for the 24 models and plotted the grids with flopy using unique colors for each model grid. It looks like there is a problem with the coordinates for the grids. It looks like some of the subdomains overlap and others are not connected. The coordinates need to be correct in order for the parallel model to define the interface models used to transfer data. 

@tandreasr could you look at your model file and make sure all of the are grids logically connected.


![tandreasr](https://github.com/user-attachments/assets/37c1f53e-4fa4-4a6f-b37d-186c9c99018f)

**Comment by tandreasr** [comment]
@jdhughes-usgs
Thanks for looking into that problem.

As requested I had a look at the subdomain coordinates.
Below you see a screenshot of our CAD based preprocessor which shows, that all subdomains are at the correct positions.
Black arrows are exchanges,
gray ones are ghostnode corrections, which are just inside the subdomains, not on their borders
(because not supported in parallel)

A problem with your floppy plot might be, that several of the subdomains use XORIGIN and YORIGIN respectively
(only explicitely set, when unequal zero),
Others however (GWF1_3_0001.disv, GWF2_1_0001.disv & GWF2_3_0001.disv)
do contain their coordinate offsets implicitely in there vertex coordinates.
But this I cannot verify because I don't use floppy.
The actual coordinates however, based on the sum of XYORIGIN (when present) and the vertex positions,
should be correct.

Best regards
Andreas

![image](https://github.com/user-attachments/assets/e543dbba-72eb-40df-89a8-3126714d86ca)

**Comment by mjr-deltares** [comment]
THanks @jdhughes-usgs and @tandreasr for the plot, that's very helpful. I was able to reproduce in debug mode on Snellius and have a clear pointer as to what the issue is. Will try and fix and report back here.

(In general, it would be helpful if the model could be written in an OS independent way. We currently only have the means to test this on a Linux system unfortunately. If you could avoid forward slashes in paths and be consistent with casing of filenames in the config files, it would save me quite a bit of time converting everything)

**Comment by mjr-deltares** [comment]
FYI your model has been tested with the updated version and ran successfully in 3.2s on 24 cores, both on a Cray machine and on an AMD supercomputer.

**Comment by tandreasr** [comment]
@mjr-deltares
works fine on my side as well :-)
Thank you very much!
Andreas
